data/src/main/java/org/apache/iceberg/data/DeleteFilter.java

```diff
@@ -226,10 +226,15 @@ private CloseableIterable<T> applyPosDeletes(CloseableIterable<T> records) {
 
     // if there are fewer deletes than a reasonable number to keep in memory, use a set
     if (posDeletes.stream().mapToLong(DeleteFile::recordCount).sum() < setFilterThreshold) {
-      return Deletes.filter(records, this::pos, Deletes.toPositionIndex(filePath, deletes));
+      Predicate<T> isInDeleteSet = record -> Deletes.toPositionIndex(filePath, deletes).isDeleted(pos(record));
```

Comment:
Won't we call `Deletes$toPositionIndex` for every row now?

Reply:
Good catch! Thanks. Fixed it in the new commit.